storage.go

```diff
@@ -2,35 +2,266 @@ package azfile
 }
 
 func (s *Storage) metadata(opt pairStorageMetadata) (meta *StorageMeta) {
-	panic("not implemented")
+	meta = NewStorageMeta()
+	meta.WorkDir = s.workDir
+	return meta
+}
+
+func (s *Storage) nextObjectPage(ctx context.Context, page *ObjectPage) error {
+	input := page.Status.(*objectPageStatus)
+
+	options := azfile.ListFilesAndDirectoriesOptions{
+		Prefix:     input.prefix,
+		MaxResults: input.maxResults,
+	}
+
+	output, err := s.client.ListFilesAndDirectoriesSegment(ctx, input.marker, options)
+	if err != nil {
+		return err
+	}
+
+	for _, v := range output.DirectoryItems {
+		o, err := s.formatDirObject(v)
+		if err != nil {
+			return err
+		}
+
+		page.Data = append(page.Data, o)
+	}
+
+	for _, v := range output.FileItems {
+		o, err := s.formatFileObject(v)
+		if err != nil {
+			return err
+		}
+
+		page.Data = append(page.Data, o)
+	}
+
+	if !output.NextMarker.NotDone() {
+		return IterateDone
+	}
+
+	input.marker = output.NextMarker
+
+	return nil
 }
 
 func (s *Storage) read(ctx context.Context, path string, w io.Writer, opt pairStorageRead) (n int64, err error) {
-	panic("not implemented")
+	offset := int64(0)
+	if opt.HasOffset {
+		offset = opt.Offset
+	}
+
+	count := int64(azfile.CountToEnd)
+	if opt.HasSize {
+		count = opt.Size
+	}
+
+	output, err := s.client.NewFileURL(path).Download(ctx, offset, count, false)
+	if err != nil {
+		return 0, err
+	}
+	defer func() {
+		cErr := output.Response().Body.Close()
+		if cErr != nil {
+			err = cErr
+		}
+	}()
+
+	rc := output.Response().Body
+	if opt.HasIoCallback {
+		rc = iowrap.CallbackReadCloser(rc, opt.IoCallback)
+	}
+
+	return io.Copy(w, rc)
 }
 
 func (s *Storage) stat(ctx context.Context, path string, opt pairStorageStat) (o *Object, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	var dirOutput *azfile.DirectoryGetPropertiesResponse
+	var fileOutput *azfile.FileGetPropertiesResponse
+
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		dirOutput, err = s.client.NewDirectoryURL(path).GetProperties(ctx)
+	} else {
+		fileOutput, err = s.client.NewFileURL(path).GetProperties(ctx)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	o = s.newObject(true)
+	o.ID = rp
+	o.Path = path
+
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		o.Mode |= ModeDir
+
+		o.SetLastModified(dirOutput.LastModified())
+
+		if v := string(dirOutput.ETag()); v != "" {
+			o.SetEtag(v)
+		}
+
+		var sm ObjectSystemMetadata
+		if v, err := strconv.ParseBool(dirOutput.IsServerEncrypted()); err == nil {
+			sm.ServerEncrypted = v
+		}
+		o.SetSystemMetadata(sm)
+	} else {
+		o.Mode |= ModeRead
+
+		o.SetContentLength(fileOutput.ContentLength())
+		o.SetLastModified(fileOutput.LastModified())
+
+		if v := string(fileOutput.ETag()); v != "" {
+			o.SetEtag(v)
+		}
+		if v := fileOutput.ContentType(); v != "" {
+			o.SetContentType(v)
+		}
+		if v := fileOutput.ContentMD5(); len(v) > 0 {
+			o.SetContentMd5(base64.StdEncoding.EncodeToString(v))
```

Comment:
ping @Xuanwo I'm not sure whether we need base64 encode here? Similar question in https://github.com/beyondstorage/go-service-azfile/pull/2/files#diff-8127f6766ba5110c7a99d07305f11d18295fcb2cb08dddb105ab7d55898742c7R254.

Reply:
Yes, our content md5 is based encoded, maybe we need to add description in `content_md5`.